File: 211.design-add-and-search-words-data-structure.py
#
# @lc app=leetcode id=211 lang=python3
#
# [211] Design Add and Search Words Data Structure
#

# @lc code=start
from collections import deque


class Trie:
    def __init__(self):
        self.tree = {}
        self.end = "#"  # use this character to indicate the end of a word

    def insert(self, word: str) -> None:
        currDict = self.tree
        w = deque(list(word))
        while w:
            letter = w.popleft()
            if letter not in currDict:
                currDict[letter] = {}
            currDict = currDict[letter]
        currDict[self.end] = True


class WordDictionary:
    def __init__(self):
        self.trie = Trie()

    def addWord(self, word: str) -> None:
        # only lowercase
        self.trie.insert(word)

    def search(self, word: str) -> bool:
        # can be lowercase and "."
        tree = self.trie.tree

        def match(currDict: dict, word: str) -> bool:
            # base case: if word == ""
            if len(word) == 0:
                return self.trie.end in currDict

            letter = word[0]
            # case 1: letter is not '.'
            if letter != ".":
                if letter not in currDict:
                    return False
                return match(currDict[letter], word[1:])
            # case 2: letter is '.'
            else:
                ansFromChildren = []
                for key in currDict:
                    # ignore the ending character
                    if key == self.trie.end:
                        continue
                    nextDict = currDict[key]
                    ansFromChildren.append(match(nextDict, word[1:]))
                return any(ansFromChildren)

        return match(tree, word)


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)
# @lc code=end

File: 211.design-add-and-search-words-data-structure.go
/*
 * @lc app=leetcode id=211 lang=golang
 *
 * [211] Design Add and Search Words Data Structure
 */

// @lc code=start
package main

type WordDictionary struct {
	t map[rune]*WordDictionary
}

func Constructor() WordDictionary {
	trie := make(map[rune]*WordDictionary)
	return WordDictionary{
		t: trie,
	}
}

func getNewDict() *WordDictionary {
	trie := make(map[rune]*WordDictionary)
	return &WordDictionary{
		t: trie,
	}
}

func (this *WordDictionary) AddWord(word string) {
	n := len(word)
	if n == 0 {
		this.t['/'] = getNewDict()
		return
	}
	firstLetter := word[0]
	rest := word[1:]
	if _, ok := this.t[rune(firstLetter)]; !ok {
		this.t[rune(firstLetter)] = getNewDict()
	}
	this.t[rune(firstLetter)].AddWord(rest)
}

func (this *WordDictionary) Search(word string) bool {
	n := len(word)
	if n == 0 {
		_, ok := this.t['/']
		return ok
	}
	firstLetter := word[0]
	rest := word[1:]
	if firstLetter == '.' {
		// iterate all children in this.t
		// use OR and return the result
		for _, v := range this.t {
			if v.Search(rest) {
				return true
			}
		}
		return false
	}

	if _, ok := this.t[rune(firstLetter)]; !ok {
		return false
	}
	return this.t[rune(firstLetter)].Search(rest)
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * obj := Constructor();
 * obj.AddWord(word);
 * param_2 := obj.Search(word);
 */
// @lc code=end
